refactor(transformer): remove unnecessary code from JSX transform

overlookmotel opened the pull request:
Follow-on after #5223.

JSX identifiers which start with a capital letter are now `JSXElementName::IdentifierReference`s, so no need to check for capitalized `JSXElementName::Identifier`s.

overlookmotel commented:
* **#5341** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/5341?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#5340** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/5340?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#5339** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/5339?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @overlookmotel and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

Dunqing commented on crates/oxc_transformer/src/react/jsx.rs:
```diff
@@ -706,11 +706,8 @@ impl<'a> ReactJsx<'a> {
             JSXElementName::Identifier(ident) => {
                 if ident.name == "this" {
                     self.ast().expression_this(ident.span)
-                } else if ident.name.chars().next().is_some_and(|c| c.is_ascii_lowercase()) {
-                    self.ast().expression_string_literal(ident.span, &ident.name)
                 } else {
```
The last remaining thing is `JSXElementName::ThisExpression`, See https://github.com/oxc-project/oxc/pull/5223#issuecomment-2320957991 😀

Dunqing reviewed (approved):
LGTM

overlookmotel commented:
Conformance is failing. This change should not have altered anything. I think that means something is wrong somewhere else - we must have ended up with a capitalized `JSXElementName::Identifier` somehow. I'll look into it.

overlookmotel commented:
I found the problem.

Identifiers starting with `_` or `$` are not being turned into `IdentiferReference`s. Weirdly, both `'_'.is_uppercase()` and `'_'.is_lowercase()` == `false`.

[Rust Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6eec69078af5f545b041ab0bd76a3ea6)

Will fix in a follow-up.

overlookmotel committed:
refactor(transformer): remove unnecessary code from JSX transform (#5339)

Follow-on after #5223.

JSX identifiers which start with a capital letter are now `JSXElementName::IdentifierReference`s, so no need to check for capitalized `JSXElementName::Identifier`s.
